### DISCUSSION TITLE
k3s check-config "optional" modules - CONFIG_CFS_BANDWIDTH

### DISCUSSION THREAD

**Discussion by parkerfath** [k3s-io/k3s#9339] [Q&A]
With k3s v1.28.5, I'm trying to set up a k3s instance on a Raspberry Pi running Debian 12 bookworm. I can get k3s installed, but once it's running, none of the containers can be created. E.g. coredns (and any other pods I try to start) gets this error:

 failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error setting cgroup config for procHooks process: openat2 /sys/fs/cgroup/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod560efa1c_48f6_44d7_a856_ed46f1305a2c.slice/cri-containerd-d16c001c17ebb1fb8e2483859bb2210e54570041611979c439f5214b7c1e522a.scope/cpu.max: no such file or directory: unknown        

From some Googling, it seems like this is possibly related to a missing CONFIG_CFS_BANDWIDTH kernel module/config. However, when I run k3s check-config, CONFIG_CFS_BANDWIDTH is under optional. But is it actually required for this case? I could try adding all the optional missing items, but just wanted to clarify my understanding of "optional" first.

FWIW, here's a summary of the missing items when I run k3s check-config:

```
Optional Features:

CONFIG_BLK_DEV_THROTTLING: missing
CONFIG_NET_CLS_CGROUP: missing
CONFIG_CGROUP_NET_PRIO: missing
CONFIG_CFS_BANDWIDTH: missing
CONFIG_RT_GROUP_SCHED: missing
CONFIG_IP_NF_TARGET_REDIRECT: missing
CONFIG_IP_SET: missing
CONFIG_IP_VS_NFCT: missing
CONFIG_IP_VS_PROTO_TCP: missing
CONFIG_IP_VS_PROTO_UDP: missing
CONFIG_IP_VS_RR: missing
Network Drivers:
"overlay":
CONFIG_VXLAN: missing
Optional (for encrypted networks):
CONFIG_CRYPTO_SEQIV: missing
CONFIG_XFRM: missing
CONFIG_XFRM_USER: missing
CONFIG_XFRM_ALGO: missing
CONFIG_INET_ESP: missing
CONFIG_INET_XFRM_MODE_TRANSPORT: missing
```
Here are the exec parts of the systemd config for k3s that I'm using:
```
ExecStartPre=/bin/sh -xc '! /usr/bin/systemctl is-enabled --quiet nm-cloud-setup.service 2>/dev/null'
ExecStartPre=-/sbin/modprobe br_netfilter
ExecStartPre=-/sbin/modprobe overlay
ExecStart=/usr/local/bin/k3s \
    server \
        '--disable=traefik' \
        '--kubelet-arg' \
        'anonymous-auth=false' \
        '--kube-apiserver-arg' \
        'allow-privileged=true' \
        '--kube-apiserver-arg' \
        'service-node-port-range=4000-5000' \
        '--flannel-backend=wireguard-native' 
```

**Comment by brandond** [comment]
You are missing a ton of kernel modules. Basically everything Kubernetes needs to run - various networking pieces, plus cgroup support.

**Reply by parkerfath** [reply]
Thanks - so if I understand correctly, these are not actually optional?

**Reply by brandond** [reply]
Which ones you need depend on which flannel backend you're using. You'll probably want all of them. The initial issue you're having appears to do with cgroup support, which won't show up in the module list checks.

**Reply by parkerfath** [reply]
thanks. FWIW I was able to get cgroup boot parameters added, though maybe there is still something missing for that?

root@myhostname:/home/myuser# cat /proc/cmdline 
root=PARTUUID=mypartitionid rootwait rootfstype=ext4 console=ttyS1,115200 console=tty1 loglevel=4 cgroup_memory=1 cgroup_enable=memory cgroup_enable=cpuset
 rw

I added the --flannel-backend=wireguard-native because before that, the node was unable to start and throwing "flannel exited: operation not supported"

If cgroup support doesn't show up in module list checks, is there some other way to check that, or a specific module/config that i need?

**Reply by brandond** [reply]
`grep cgroup /proc/mounts` and `cat /proc/cgroups` are usually what I look at.

**Reply by parkerfath** [reply]
Thanks. This is what I found:
```
root@host:/home/user# grep cgroup /proc/mounts
cgroup2 /sys/fs/cgroup cgroup2 rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot 0 0
root@host:/home/user# cat /proc/cgroups
#subsys_name    hierarchy       num_cgroups     enabled
cpuset  0       352     1
cpu     0       352     1
cpuacct 0       352     1
blkio   0       352     1
memory  0       352     1
devices 0       352     1
freezer 0       352     1
perf_event      0       352     1
hugetlb 0       352     1
pids    0       352     1
```

**Reply by parkerfath** [reply]
OK thanks again; I was able to get around to adding more of the optional kernel modules, and k3s runs a *little* better now. However, I'm still having some issues with k3s restarting after installing some of my own custom deployments/pods.

1. When I just install "plain" k3s with disable=traefik and add my own nginx-ingress, I'm fine.
2. However, when I start to add other pods, e.g. postgres or redis, I start getting k3s restarts every 5 minutes.

When I look in the k3s journalctl logs, I don't see much that looks like an obvious reason for the restart. 

Before k3s shuts down, I see:
level=info msg="Shutting down /v1, Kind=ConfigMap workers"
Stopping k3s.service - Lightweight Kubernetes...

And I also noticed higher up in the logs, 

network_policy_controller.go:311] Aborting sync. Failed to run iptables-restore: failed to call iptables-restore:>
Warning: Extension NFLOG revision 0 not supported, missing kernel module?
iptables-restore: line 281 failed

Note that this was with using iptables-legacy and ip6tables-legacy, though I saw pretty similar behavior on iptables-nftables.

In check-config, these are the modules I'm still seeing missing:
Optional Features:
- CONFIG_BLK_DEV_THROTTLING: missing
- CONFIG_CGROUP_NET_PRIO: missing
- CONFIG_RT_GROUP_SCHED: missing
- CONFIG_IP_VS_NFCT: missing
      - CONFIG_CRYPTO_SEQIV: missing
      - CONFIG_XFRM: missing
      - CONFIG_XFRM_USER: missing
      - CONFIG_XFRM_ALGO: missing
      - CONFIG_INET_ESP: missing
      - CONFIG_INET_XFRM_MODE_TRANSPORT: missing

So... do I need to add something for NFLOG? Or is there another place I can look for more descriptive error logs? Thanks again!

**Reply by brandond** [reply]
> ```
> level=info msg="Shutting down /v1, Kind=ConfigMap workers"
> Stopping k3s.service - Lightweight Kubernetes...
> ```
Something else is stopping the k3s systemd service.. it's not crashing. You should figure out what that is.

**Reply by parkerfath** [reply]
Yes, you're totally right. Thanks for pointing that out. :) 

We were able to get things running well with the above-mentioned changes in the kernel modules, and the cluster seems pretty stable so far.

I would turn it back around, though, to the original question: would it make sense to call out more of these modules as required rather than optional? It could be helpful for future people who run into the same problem.
